### PR TITLE
Browser Hotkey Conflicts

### PR DESCRIPTION
**Pull request by thomas992**
### EasyMark Text Editor

**Hotkey Limitations Within Browser**
ctrl A S D F G H J K are all taken the entire second row besides L. CTRL W E R T U O P are also all taken, entire first row besides QYI. CTRL Z X C V are taken so we are left with a limited number of keys for each row, BNM in third.

LQYIBNM, 7 out of the 26 letters are all the keys we have left.

**New Hotkey Combinations**

as the project is unfinished I added ALTSHIFT modifier to enable more hotkey variations. The keys for , . ; are missing with the semicolon being noted as of not being internationally according to comment. I think we need to be able to access these keys in the future through an international keyboard interpreting algorithm.

**New view**

I thought that context menus would be more appropriate to move forward with. I want the hotkeys thing to bring up a window for editing hotkeys along with viewing what we think are the system hotkeys as well.

caption
Closes <https://github.com/emilk/egui/issues/1683>.

### CODE REVIEW

**Comment by thomas992** [comment]
updated text and moved strikethrough CTRL+S to CTRL+M.

caption

this is how Google Docs is set up, strikethrough is a hotkey you have to look for.

**Comment by thomas992** [comment]
The ALT shift commands could be a great alternative to these made up hotkeys imo

**Comment by thomas992** [comment]
I just tested the tab key does not actually work on browser that is whole issue in itself

**Comment by thomas992** [comment]
The EasyMark editor assumes that the keyboard is capable of pressing / ` $ ^ ~ _ but this is not the case internationally.

**Comment by thomas992** [comment]
Would using

```rust
use std::env;

        if env::consts::OS == "macos" || env::consts::OS == "ios" {
            ui.label("CMD + [B: *bold*  N: `code`  I: /italics/  L: $lower$  Y: ^raise^  M: ~Strikethrough~] ALTSHIFT+Q: _underline_");
        } else {
            ui.label("CTRL + [B: *bold*  N: `code`  I: /italics/  L: $lower$  Y: ^raise^  M: ~Strikethrough~] ALTSHIFT+Q: _underline_");
        }
```
be acceptable? it seems like not the right way I seen other places in the code that does like a cfg target thing I guess that would be the right way

**Comment by thomas992** [comment]
ALTSHIFT is now treated as a command, all the hotkeys are fully working now. The last two things are printing OS appropriate hotkeys and adding indentations.

**Comment by thomas992** [comment]
ALTSHIFT Num1 is not registering for me.

**Comment by thomas992** [comment]
caption
I am looking into maintaining cursor range while navigating context menu I am not sure if that is feasible

**Comment by thomas992** [comment]
#1705 I would rather continue the EasyMark editor once I implement some sort of Rhai support. I think Rhai files should be standard in EGUI for widgets and windows. Hotkeys are implemented in a way that do not support an integrated development environment. Operating a web assembly website the greatest advantage is run time compilation.

File: eframe/src/web/events.rs [hunk 1]
```diff
@@ -50,7 +50,7 @@ pub fn install_document_events(runner_container: &AppRunnerContainer) -> Result<
         "keydown",
         |event: web_sys::KeyboardEvent, mut runner_lock| {
             if event.is_composing() || event.key_code() == 229 {
-                // https://www.fxsitecompat.dev/en-CA/docs/2018/keydown-and-keyup-events-are-now-fired-during-ime-composition/
+                // https://web.archive.org/web/20200526195704/https://www.fxsitecompat.dev/en-CA/docs/2018/keydown-and-keyup-events-are-now-fired-during-ime-composition/
```
Comment (emilk) [inline eframe/src/web/events.rs:53]: 👍

File: egui_demo_lib/src/easy_mark/easy_mark_editor.rs [hunk 2]
```diff
@@ -45,15 +45,19 @@ impl EasyMarkEditor {
 
     pub fn ui(&mut self, ui: &mut egui::Ui) {
         egui::Grid::new("controls").show(ui, |ui| {
+            //ui.menu_button("File", Self::nested_file);
+            ui.menu_button("Hotkeys", Self::nested_hotkeys);
+            //ui.menu_button("View", Self::nested_view);
```
Comment (emilk) [inline egui_demo_lib/src/easy_mark/easy_mark_editor.rs:50]:
```suggestion
            ui.menu_button("Hotkeys", Self::nested_hotkeys);
```

File: egui_demo_lib/src/easy_mark/easy_mark_editor.rs [hunk 3]
```diff
@@ -45,15 +45,19 @@ impl EasyMarkEditor {
...
+            if ui
+                .button(RichText::new("render").color(Color32::from_rgb(76, 139, 255)))
+                .clicked()
+            {
+                self.show_rendered = !self.show_rendered;
```
Comment (emilk) [inline egui_demo_lib/src/easy_mark/easy_mark_editor.rs:55]: this should `ui.checkbox` or `ui.toggle_value` (like it was)

File: egui_demo_lib/src/easy_mark/easy_mark_editor.rs [hunk 4]
```diff
@@ -107,22 +111,52 @@ impl EasyMarkEditor {
...
     }
+    fn nested_hotkeys(ui: &mut egui::Ui) {
+        let _ = ui.button("CTRL+B *bold*");
+        let _ = ui.button("CTRL+N `code`");
+        let _ = ui.button("CTRL+I /italics/");
+        let _ = ui.button("CTRL+L $subscript$");
+        let _ = ui.button("CTRL+Y ^superscript^");
+        let _ = ui.button("CTRL+M ~strikethrough~");
+        let _ = ui.button("ALTSHIFT+Q _underline_");
+        let _ = ui.button("ALTSHIFT+W two spaces");
```
Comment (emilk) [inline egui_demo_lib/src/easy_mark/easy_mark_editor.rs:122]:
```suggestion
        let _ = ui.button("ALT+SHIFT+Q _underline_");
        let _ = ui.button("ALT+SHIFT+W two spaces");
```

File: egui_demo_lib/src/easy_mark/easy_mark_editor.rs [hunk 5]
```diff
@@ -107,22 +111,52 @@ impl EasyMarkEditor {
             }
         }
     }
+    fn nested_hotkeys(ui: &mut egui::Ui) {
+        let _ = ui.button("CTRL+B *bold*");
```
Comment (emilk) [inline egui_demo_lib/src/easy_mark/easy_mark_editor.rs:115]: Why buttons if they don't do anything? Either make them labels, or make them do what they say they do!

File: egui_demo_lib/src/easy_mark/easy_mark_editor.rs [hunk 6]
```diff
@@ -107,22 +111,52 @@ impl EasyMarkEditor {
...
+    }
+    fn nested_file(ui: &mut egui::Ui) {
+        let _ = ui.button("CTRL+O Open");
+        let _ = ui.button("New File");
+        //let _ = ui.button("Reset");
+    }
+    /*fn nested_view(ui: &mut egui::Ui) {
+        let _ = ui.button("Highlight Editor");
+        let _ = ui.button("Show Rendered");
+    }*/
```
Comment (emilk) [inline egui_demo_lib/src/easy_mark/easy_mark_editor.rs:132]: rm dead code

File: egui/src/data/input.rs [hunk 7]
```diff
@@ -323,6 +323,13 @@ impl Modifiers {
         mac_cmd: false,
         command: false,
     };
+    pub const ALTSHIFT: Self = Self {
```
Comment (emilk) [inline egui/src/data/input.rs:326]:
```suggestion
    pub const ALT_SHIFT: Self = Self {
```

**Review by emilk — CHANGES_REQUESTED** [review]

**Review by thomas992 — COMMENTED** [review]
update

File: egui_demo_lib/src/easy_mark/easy_mark_editor.rs [hunk 8]
```diff
@@ -107,22 +104,43 @@ impl EasyMarkEditor {
...
     }
+    fn nested_hotkeys(ui: &mut egui::Ui) {
+        let _ = ui.label("CTRL+B *bold*");
+        let _ = ui.label("CTRL+N `code`");
+        let _ = ui.label("CTRL+I /italics/");
+        let _ = ui.label("CTRL+L $subscript$");
+        let _ = ui.label("CTRL+Y ^superscript^");
+        let _ = ui.label("ALT+SHIFT+Q ~strikethrough~");
+        let _ = ui.label("ALT+SHIFT+W _underline_");
+        let _ = ui.label("ALT+SHIFT+E two spaces");
```
Comment (emilk) [inline egui_demo_lib/src/easy_mark/easy_mark_editor.rs:115]: Wait I missed  this in my first review - what weird shortcut is this? O.o

Comment (thomas992) [reply egui_demo_lib/src/easy_mark/easy_mark_editor.rs:115]: the two spaces is a placeholder for tab indenting

**Comment by thomas992** [comment]
I have not tested this on WASM

**Review by thomas992 — COMMENTED** [review]
review

**Review by emilk — APPROVED** [review]

**Pull request merged**